File: Olympiad/IOI/official/2017/practice-problems/cup/cup.java
import java.io.*;
import java.util.*;


class cup{
	final int maxn = 500000000;
	int x,tmp,tmp1,y;

	boolean same(int k) {
		long help=((1L)<<33)-1-((1L)<<k);
		return ((tmp&help)==(tmp1&help));
	}

	long f(long k) {
		for(int i=33;i>-1;i--)
			if(((k>>i)&1L) != 0L)
				return ((1L)<<i);
		return 0L;
	}

	public int[] find_cup() {
		int[] result = new int[2];

		x=y=tmp1=0;
		tmp=grader.ask_shahrasb(-2*maxn,-maxn);
		for(int i=0;i<31;i++)
		{
			long k=((1L)<<i);
			tmp1=grader.ask_shahrasb((int)(-2*maxn+k),Integer.valueOf(-maxn));
			if(same(i))
				y+=k^(tmp&k);
			else
				y+=(tmp&k);
		}
		x=y^tmp;
		if(-2*maxn+x>maxn)
			x-=f(x);
		result[0] = Integer.valueOf(-2*maxn+x);
		result[1] = Integer.valueOf(-maxn+(tmp^x));
		return result;
	}
}

File: Olympiad/IOI/official/2017/practice-problems/cup/cup.pas
unit cup;

interface
uses graderlib;

function find_cup(): TIntArray;

implementation

const
    maxn: longint = 500 * 1000 * 1000;
    one: int64 = 1;

var
    x, tmp, tmp1, y: int64;

function same(k: int64): boolean;
var
    help: int64;
begin
    help := (one shl 33) - one - (one shl k);
    same := (tmp and help) = (tmp1 and help);
end;

function f(k: int64): int64;
var
    i: longint;
begin
    for i := 33 downto 0 do
    begin
        if ((k shr i) and 1) = 1 then
        begin
            f := one shl i;
            exit;
        end;
    end;
    f := 0;
end;

function find_cup(): TIntArray;
var
    i: longint;
    k: int64;
    result: TIntArray;
begin
    x := 0;
    y := 0;
    tmp1 := 0;
    tmp := ask_shahrasb(-2 * maxn, -maxn);

    for i := 0 to 30 do
    begin
        k := one shl i;
        tmp1 := ask_shahrasb(-2 * maxn + k, -maxn);
        if same(i) then
            y := y + (k xor (tmp and k))
        else
            y := y + (tmp and k);
    end;

    x := y xor tmp;

    if -2 * maxn + x > maxn then
        x := x - f(x);
        
    setlength(result, 2);
    result[0] := -2 * maxn + x;
    result[1] := -maxn + (tmp xor x);
    find_cup := result;
end;

end.
